### PR TITLE
Completed Project 3 - Part 2 - for review

### PR DESCRIPTION
**Pull request by bbgreene**

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -121,385 +163,563 @@ struct CarWash
...
+        int personalEmailTotal; // added member variable for nested 2nd implementation
+        int programmesOpen; // added member variable for nested 3rd implementation
 		
-		bool turnOnLaptop(bool laptopOff = true, bool userPressedPowerButton = true);
-		int sendEmails(int emailsPerDay = 10, bool personalEmails = true);
-		void openAProgram(int currentProgramAmount, bool isUserAdmin);
+		void turnOnLaptop(bool laptopOff, bool userPressedPowerButton);
+		auto sendEmails(bool personalEmails = true, int emailsPerDay = 10);
+		auto openAProgram(int currentProgramAmount, bool newProgrammeOpen);
 	};
```
Comment (matkatmusic) [inline main.cpp]: You've got some weird indentation here.

File: main.cpp [hunk 2]
```diff
@@ -121,385 +163,563 @@ struct CarWash
...
+ if(personalEmails == true)
+ {
+    emailsPerDay = 10;
+ }
+ else
+ {
+    emailsPerDay = 0;
+ }
+personalEmailTotal += emailsPerDay;
+}
```
Comment (matkatmusic) [inline main.cpp]: Fix the indentation
- [x] revise

File: main.cpp [hunk 3]
```diff
@@ -121,385 +163,563 @@ struct CarWash
...
+
+void Laptop::automaticallyAdjustBrightness(float timeOfDay, int displayBrightnessLevel)
+{
+    if(timeOfDay > 6.0f && timeOfDay < 18.0f)
+    {
+        displayBrightnessLevel = 80;
+    }
+    else
+        displayBrightnessLevel = 35;
+}
```
Comment (matkatmusic) [inline main.cpp]: If you're going to use `{ }`, use it for both `if` and `else` blocks.
- [x] revise

File: main.cpp [hunk 4]
```diff
@@ -121,385 +163,563 @@ struct CarWash
...
 
-	//output MIDI information
-	int outputMIDIInfo(int midiNoteValue = 62, int velocityValue = 71);
-	//display MIDI note value
-	void diplayMIDINoteValue(bool isKeyPressed);
-	//consume electricity via USB port
-	float consumeElectricity(bool pluggedInToHost = true, float wattage = 10.4f);
+	void outputMIDIInfo(int midiNoteValue, int velocityValue);
+	void diplayMIDINoteValue(bool isKeyPressed, int midiNoteValue);
+	void consumeElectricity(bool pluggedInToHost, float wattage);
```
Comment (matkatmusic) [inline main.cpp]: Fix the indentation
- [x] revise

File: main.cpp [hunk 5]
```diff
@@ -121,385 +163,563 @@ struct CarWash
...
+    float outputLevel; // new member variable for implementation 3 below
 
-	//generate audio signals
-	void generateAudioSignals(float oscFreq, int waveType);
-	//automatically adjust filter frequency
+	void generateAudioSignals(float maxOscFreq, int waveType);
 	void autoFilterAdjust(int freqRange, bool useLFO, bool useEnv1); 
-	//output audio signals
-	float outputAudioSignals(bool stereo = true, float maxOutput = 24.0f);
+	void outputAudioSignals(bool keyHasBeenPressed, float maxOutput);
```
Comment (matkatmusic) [inline main.cpp]: Fix the indentation
- [x] revise

File: main.cpp [hunk 6]
```diff
@@ -121,385 +163,563 @@ struct CarWash
...
+    if(keyHasBeenPressed == true)
+    {
+        output = 24.0f;
+    }
+   else 
+   {
+        output = 0.0f;
+   }
+   outputLevel += output;
+}
```
Comment (matkatmusic) [inline main.cpp]: Fix the indentation
- [x] revise

File: main.cpp [hunk 7]
```diff
@@ -121,385 +163,563 @@ struct CarWash
...
+{
+    if(changePitch > 0)
+    {
+        transpose += changePitch;
+    }
+    else 
+    {
+        transpose -= changePitch;
+    }
+}
```
Comment (matkatmusic) [inline main.cpp]:
```
{
   transpose += changePitch;
}
```
if `changePitch` is negative, it'll be subtracted from `transpose`.  no need for the `if/else`
- [x] revise

File: main.cpp [hunk 8]
```diff
@@ -121,385 +163,563 @@ struct CarWash
...
+    if(feedbackValue > 0.7f)
+    {
+        selfOscAmount += feedbackValue;
+    }
+}
+
+void Filter::CircuitTypeEmulator::clipAudio(std::string clipType)
+{
+    clipType = "Soft";
+}
```
Comment (matkatmusic) [inline main.cpp]: You are overwriting your function parameter with `"Soft"`.  What is the point of your function parameter if you're not even using it?
You need a member variable that `clipType` can be assigned to.
- [x] revise

File: main.cpp [hunk 9]
```diff
@@ -121,385 +163,563 @@ struct CarWash
...
-    1) modulate amplitude of sample 
-    2) modulate pitch of sample
-    3) modulate audio output between left and right channel
- */
+void Envelope::controlAmpOverTime(bool ampEnvSelected, bool isKeyPressed, float maxValueAmp)
+{
+    if(isKeyPressed == true && ampEnvSelected == true)
+    {
+        maxValueAmp = midiVelLevel / 100;
+    }
```
Comment (matkatmusic) [inline main.cpp]: You don't need to compare these booleans with `true`.  `if( isKeyPressed && ampEnvSelected )` will do the trick
- [x] revise

File: main.cpp [hunk 10]
```diff
@@ -121,385 +163,563 @@ struct CarWash
...
+        maxValueAmp = 0.0f;
+    }
+}
+
+void Envelope::controlCutoffOverTime(bool filterEnvSelected, bool isKeyPressed, double maxValueFilter)
+{
+    if(isKeyPressed == true && filterEnvSelected == true)
+    {
+        maxValueFilter = midiVelLevel * 10;
+    }
```
Comment (matkatmusic) [inline main.cpp]: no need for `== true`
- [x] revise

File: main.cpp [hunk 11]
```diff
@@ -121,385 +163,563 @@ struct CarWash
...
+    else 
+    {
+        maxValueFilter = 1000;
+    }
+}
+
+void Envelope::controlPitchOverTime(bool pitchEnvSelected, bool isKeyPressed, double maxValuePitch)
+{
+    if(isKeyPressed == true && pitchEnvSelected == true)
```
Comment (matkatmusic) [inline main.cpp]: no need for `== true`
- [x] revise

File: main.cpp [hunk 12]
```diff
@@ -121,385 +163,563 @@ struct CarWash
...
-    4) Envelope
-    5) LFO
-3 things it can do:
-    1) display waveform of an audio sample 
-    2) playback audio at various speeds
-    3) effect frequency content of audio
- */
+void LFO::modAmpOfSample(bool lfoON, int ampRange, bool ampToggle = true)
+{
+    if(lfoON == true && ampToggle == true)
```
Comment (matkatmusic) [inline main.cpp]: - [x] revise

File: main.cpp [hunk 13]
```diff
@@ -121,385 +163,563 @@ struct CarWash
...
+    else
+    {
+        ampRange = 0;
+    }
+    modDest += 0;
+}
+
+void LFO::modPitchOfSample(bool lfoON, int pitchRange, bool pitchToggle = true)
+{
+    if(lfoON == true && pitchToggle == true)
```
Comment (matkatmusic) [inline main.cpp]: - [x] revise

File: main.cpp [hunk 14]
```diff
@@ -121,385 +163,563 @@ struct CarWash
...
+    else
+    {
+        pitchRange = 0;
+    }
+    modDest += 1;
+}
+
+void LFO::modPanOfSample(bool lfoON, int panRange, bool panToggle = true)
+{
+    if(lfoON == true && panToggle == true)
```
Comment (matkatmusic) [inline main.cpp]: - [x] revise

File: main.cpp [hunk 15]
```diff
@@ -121,385 +163,563 @@ struct CarWash
...
-    //affect frequency content of audio
-    void freqContentOverall();
+    void displayVisualWaveform(bool isSampleLoaded, bool displayAWaveform);
+    void playbackSpeed(bool warpOff, bool normalSpeed);
+    void playbackSamples(bool stereoSampleLoaded);
 };
+
+void VSTSampler::displayVisualWaveform(bool isSampleLoaded, bool displayAWaveform)
+{
+    if(isSampleLoaded == true)
```
Comment (matkatmusic) [inline main.cpp]: - [x] revise

File: main.cpp [hunk 16]
```diff
@@ -121,385 +163,563 @@ struct CarWash
...
+	struct User
 	{
 		bool isUserMale = true;
 		int userAge = 37;
-		bool isUserAdmin = true;
+		bool isUserAdmin;
 		std::string userName = "Billy Crystal";
 		std::string password = "123456789";
+        int personalEmailTotal; // added member variable for nested 2nd implementation
+        int programmesOpen; // added member variable for nested 3rd implementation
```
Comment (matkatmusic) [inline main.cpp]: "for nested 2nd implementation"? 
what are you talking about?

**Review by matkatmusic — CHANGES_REQUESTED** [review]
- You are comparing a lot of booleans with `true`, which you don't need to do.
- you are also overwriting your function parameters inside the functions. 
create member variables that your function parameters are assigned to or modify.
- [x] revise

File: main.cpp (hunk 1 above)
Comment (bbgreene) [reply main.cpp]: weird indeed. Doesn't appear like this in replit. It seems to be common to any variables with comments beside them. I will delete the comments and see

File: main.cpp (hunk 16 above)
Comment (bbgreene) [reply main.cpp]: I made some new member variables so that the implementations below had something to return values to.

Comment (bbgreene) [reply main.cpp]: like your "distanceTraveled" example in the video for this project part

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**